### ISSUE TITLE
Feature Request: tooltips when hovering on widgets

### ISSUE THREAD

**Issue by pawamoy** [qtile/qtile#1441]
It could be nice to have tooltips showing up when hovering on a widget, be it to display it's mouse bindings or display some information. The text could be static, or dynamic by providing a command from which to get the text. The command could be run at custom intervals. A bit like what Conky offers :slightly_smiling_face:

**Comment by zordsdavini** [comment]
Maybe it can be not as tooltip but change widget text. Example, I have battery widget and display only left percentage but with this hover I can display all information (remaining time, watts etc. as it is in default text). Another example Gmail widget (actually now it is not working for me), I used it to display only not read emails, but with hover it could give full information: total emails and unread.
Additionally it could get some different color

**Comment by pawamoy** [comment]
Hmm, this is interesting as well! And maybe easier to implement (maybe not:  a tooltip would be an addition to the code, whereas changing the text of the widget would require changing the widgets code).

In fact, if #1440 gets implemented, we could have fake tooltips using notifications. Example: when clicking on your Gmail widget, a notification (using your system's notifications) would pop, showing you extra info :slightly_smiling_face:

**Comment by zordsdavini** [comment]
hovering and pushing notifications would give a lot off noise but maybe this could be the option. Even I imagine it could be possible to expand top bar with text bellow temporally (just brainstorming)

**Comment by m-col** [comment]
(How) Is it possible to listen to mouse-enter events at the level of individual widgets?

**Comment by tych0** [comment]
I think you'd need something like, https://github.com/tych0/qtile/commit/c01d4245557918e68cb2ee4a4716000226e5b839

**Comment by m-col** [comment]
Correct me if I'm wrong, but wouldn't that only pass entry/leaving of the entire bar window, but not entry/leave of a widget to/from another widget?

**Comment by tych0** [comment]
Yeah, true. So you'll need the above plus some other code handling MotionNotify too.

**Comment by ramnes** [comment]
Unless each widget is a client? Not sure if that's feasible or if that makes sense though.

**Comment by tych0** [comment]
I think it's probably simpler to just handle MotionNotify.

**Comment by m-col** [comment]
I've got the window side of things working on [this branch](https://github.com/m-col/qtile/tree/widget_tooltips).

In that commit the tooltip dimensions are the same as the widget's or configured by passing `tooltip_size=[width, height]` to the widget. I added it to the `Battery` widget as a proof of concept, but they'll have to hook into the widget's polling or whatever to get the correct information to display (I just stuck in the string `"100%"`). Here's a demo:

![widget_tooltips_demo](https://user-images.githubusercontent.com/44536113/88456963-245eaf00-ce72-11ea-9968-9037c08743f3.gif)

Updating the tooltip's size is just a matter of the widget setting `self.tooltip.width` and `self.tooltip.height` and then calling `self.tooltip.place()`* but something would have to calculate the size based on what is to be displayed. I'm not sure what the best approach would be for that. Maybe tooltips could actually be widgets themselves, just using a different window?

Any thoughts?

\* this could probably be simplified by allowing `Popup.place(width=width, height=height)`.

**Comment by elParaguayo** [comment]
This is still on my To Do list: (https://github.com/qtile/qtile/pull/2249 was a start...)
